### PR TITLE
[P1-L03] MultiRole roles can be renounced

### PR DESCRIPTION
**Pull request by nicholaspai**
This PR adds a feature allowing users to renounce their `MultiRole` SharedRole memberships, which could be useful if role holder accounts get compromised.

Currently, the only users who can remove role membership from other accounts must hold the SharedRole Managing role.

I added a method `renounceMembership` that allows someone to renounce their `SharedRole` membership even if they do not hold the managing role. 

After discussion with the team off-line, we decided not to allow users to renounce their ExclusiveRole membership because doing so accidentally could cause ownership of that role to be lost forever. We deemed that this downside risk was not worth the upside. One example of this risk would be in the `DesignatedVoting` contract which assigns an "Owner" role (an ExclusiveRole) capable of withdrawing funds and changing the Voter account. If the Owner were to accidentally renounce their membership, funds would be permanently locked in the contract.

There is no equivalent danger with a `SharedRole` because the role holders ideally set up backup member accounts such that the role would never get to a state in which no one was a role holder at any given moment.

Longer term, we'd like to deprecate ExclusiveRoles in favor of safer and more flexible SharedRoles, but the main obstacle from doing so is that certain already-deployed contracts like the `DesignatedVoting` contract already give `ExclusiveRoles` privilege to transfer funds.

Signed-off-by: Nick Pai <npai.nyc@gmail.com>

### CODE REVIEW

**Comment by chrismaree** [comment]
Cool that you can already renounce your roll but you should be able to renounce an `ExclusiveRole` I think. To counter the downside you mention could we not enforce that this kind or role can only be renounced iff you have transferred it away first such that it cant get lost going forward?

**Comment by nicholaspai** [comment]
> Cool that you can already renounce your roll but you should be able to renounce an `ExclusiveRole` I think. To counter the downside you mention could we not enforce that this kind or role can only be renounced iff you have transferred it away first such that it cant get lost going forward?

How is that different from calling `resetMember`? Remember that an `ExclusiveRole` can only be held by one account at a time. And resetting membership can only be called by that member.

File: core/test/common/MultiRole.js [hunk 1]
```diff
@@ -137,6 +142,11 @@ contract("MultiRole", function(accounts) {
     await multiRole.removeMember("2", account2, { from: account1 });
     assert.isFalse(await multiRole.holdsRole("2", account2));
     assert(await didContractThrow(multiRole.revertIfNotHoldingRole("2", { from: account2 })));
+
+    // Anyone who who holds the managing role can renounce their membership.
```
Comment (mrice32) [inline core/test/common/MultiRole.js:146]: I think `L03` is referring to the case where the role holder *does not* hold the managing role. Renouncing a role is basically a special case where the holder of a role should be able to give up their own membership, even if they don't have the power to modify anyone else's membership (or add themselves back). Does that make sense? Or do you think I'm misinterpreting the OZ issue?

Comment (nicholaspai) [reply core/test/common/MultiRole.js:146]: I think this special case is correct. An `ExclusiveRole` must have a separate RoleManager from a RoleHolder. In this situation, they are recommending we add an ability for the RoleHolder to renounce their membership.

Comment (nicholaspai) [reply core/test/common/MultiRole.js:146]: For `SharedRoles` the caller who wants to renounce their membership can just call `removeMember` and pass in their own address. So I think we should just add a function renounces membership specifically for ExclusiveRoles

Comment (nicholaspai) [reply core/test/common/MultiRole.js]: I made the changes we've discussed in this thread now

Comment (mrice32) [reply core/test/common/MultiRole.js]: Wait, are you sure about the SharedRole?

Let's say `address1` is a member of `role1` (exclusive) and `address2` is a member of `role2` (shared). `role1` manages `role2`. My understanding is that `address2` cannot remove itself from `role2`. `address1` is the only one with permission to do that. Is that not the case?

Comment (nicholaspai) [reply core/test/common/MultiRole.js]: @mrice32 Ah yes, I didn't word that correctly. So we actually should add a way for non-managers to renounce their shared role membership

Comment (nicholaspai) [reply core/test/common/MultiRole.js]: Ok I fixed this. Now both exclusive and shared roles can be renounced by any holder via `renounceMembership`

**Comment by mrice32** [comment]
@ptare could you take a look at this as well. This is a fairly sensitive change, so I want to make sure we get extra eyes on it.

**Comment by ptare** [comment]
The PR itself looks good to me. Are we concerned about any scenarios where we could bork the deployment by renouncing roles by mistake (especially for the self managed exclusive roles)? 

Also, IIUC contract creators can now deregister themselves and the voter in DesignatedVoting can unilaterally renounce its power. Both seem fine to me, just wanted to explicitly call out.

Devil's advocate: we don't really have a need for this feature in our system (if a role is compromised, the owner can remove it anyway) and it _might_ increase the attack surface or possibility of mistakes. Down the road, we'll migrate to a more fully features access control framework that OZ has built.

**Comment by mrice32** [comment]
> The PR itself looks good to me. Are we concerned about any scenarios where we could bork the deployment by renouncing roles by mistake (especially for the self managed exclusive roles)?
> 
> Also, IIUC contract creators can now deregister themselves and the voter in DesignatedVoting can unilaterally renounce its power. Both seem fine to me, just wanted to explicitly call out.
> 
> Devil's advocate: we don't really have a need for this feature in our system (if a role is compromised, the owner can remove it anyway) and it _might_ increase the attack surface or possibility of mistakes. Down the road, we'll migrate to a more fully features access control framework that OZ has built.

@nicholaspai @chrismaree thoughts?

**Comment by nicholaspai** [comment]
> > The PR itself looks good to me. Are we concerned about any scenarios where we could bork the deployment by renouncing roles by mistake (especially for the self managed exclusive roles)?
> > Also, IIUC contract creators can now deregister themselves and the voter in DesignatedVoting can unilaterally renounce its power. Both seem fine to me, just wanted to explicitly call out.
> > Devil's advocate: we don't really have a need for this feature in our system (if a role is compromised, the owner can remove it anyway) and it _might_ increase the attack surface or possibility of mistakes. Down the road, we'll migrate to a more fully features access control framework that OZ has built.
> 
> @nicholaspai @chrismaree thoughts?

Great point about DesignatedVoting. I am a bit anxious that someone could unintentionally renounce their ownership of that contract and get funds locked there permanently.

**Comment by nicholaspai** [comment]
@ptare @chrismaree @mrice32 I'm all for enabling users to renounce their `SharedRole` membership, but I'm hesitant about renouncing the `ExclusiveRole` membership as this could potentially get the Role in an irrecoverable state (i.e. no one can become a member ever again).

The counter-argument to this is that even if a user cannot renounce their `ExclusiveRole`, they just as easily (?) could reset their membership to some uncontrolled key (via fat finger, or maliciously) after which, ownership could never get reclaimed.

Ideally, the ExclusiveRole membership owner would be some sort of MultiSig group...

Maybe the best solution is to document well these dangers and make sure that all accounts that own important ExclusiveRoles (such as the Owner in a DesignatedVoting contract) are tied to very secure keys, such as hardware wallets.

**Comment by chrismaree** [comment]
> @ptare @chrismaree @mrice32 I'm all for enabling users to renounce their `SharedRole` membership, but I'm hesitant about renouncing the `ExclusiveRole` membership as this could potentially get the Role in an irrecoverable state (i.e. no one can become a member ever again).
> 
> The counter-argument to this is that even if a user cannot renounce their `ExclusiveRole`, they just as easily (?) could reset their membership to some uncontrolled key (via fat finger, or maliciously) after which, ownership could never get reclaimed.
> 
> Ideally, the ExclusiveRole membership owner would be some sort of MultiSig group...
> 
> Maybe the best solution is to document well these dangers and make sure that all accounts that own important ExclusiveRoles (such as the Owner in a DesignatedVoting contract) are tied to very secure keys, such as hardware wallets.

Ye. I agree with this hesitancy here. The situation in which the only exclusive role renounces their position and we have perpetually lost access to recover that state is a catastrophic system failure at this point. This should be avoided at all costs, even if this means limiting how role renouncement works for Exclusive roles. 

Could we place logic in the MultiRole to deal with this kind of thing? for example an exclusive role cant be renounced if their is only one address holding the role, for example.

**Comment by nicholaspai** [comment]
> > @ptare @chrismaree @mrice32 I'm all for enabling users to renounce their `SharedRole` membership, but I'm hesitant about renouncing the `ExclusiveRole` membership as this could potentially get the Role in an irrecoverable state (i.e. no one can become a member ever again).
> > The counter-argument to this is that even if a user cannot renounce their `ExclusiveRole`, they just as easily (?) could reset their membership to some uncontrolled key (via fat finger, or maliciously) after which, ownership could never get reclaimed.
> > Ideally, the ExclusiveRole membership owner would be some sort of MultiSig group...
> > Maybe the best solution is to document well these dangers and make sure that all accounts that own important ExclusiveRoles (such as the Owner in a DesignatedVoting contract) are tied to very secure keys, such as hardware wallets.
> 
> Ye. I agree with this hesitancy here. The situation in which the only exclusive role renounces their position and we have perpetually lost access to recover that state is a catastrophic system failure at this point. This should be avoided at all costs, even if this means limiting how role renouncement works for Exclusive roles.
> 
> Could we place logic in the MultiRole to deal with this kind of thing? for example an exclusive role cant be renounced if their is only one address holding the role, for example.

@chrismaree The problem is that by definition, the `ExclusiveRole` is only held by one account 😏

**Pull request merged**